documentation/docs/05-misc/02-testing.md

```diff
@@ -2,7 +2,156 @@
+
+### Using runes inside your test files
+
+It is possible to use runes inside your test files. First ensure your bundler knows to route the file through the Svelte compiler before running the test by adding `.svelte` to the filename (e.g `multiplier.svelte.test.js`). After that, you can use runes inside your tests.
+
+```js
+/// file: counter.svelte.test.js
+import { flushSync, mount, unmount } from 'svelte';
+import { expect, test } from 'vitest';
+import { multiplier } from './multiplier.svelte.js';
+
+test('Multiplier', () => {
+	let count = $state(0);
+	let double = multiplier(() => count, 2);
+
+	expect(double.value).toEqual(0);
+
+	count = 5;
+
+	expect(double.value).toEqual(10);
+});
+```
+
+If the tested code uses effects, you need to wrap the test inside `$effect.root` to create a scope in which these are properly captured.
+
+```js
+/// file: logger.svelte.test.js
+import { flushSync, mount, unmount } from 'svelte';
+import { expect, test } from 'vitest';
+import { logger } from './logger.svelte.js';
+
+test('Effect', () => {
+	const cleanup = $effect.root(() => {
+		let count = $state(0);
+		// logger uses an $effect to log updates of its input
+		let log = logger(() => count);
+
+		// effects normally run after a microtask,
+		// use flushSync to execute all pending effects synchronously
+		flushSync();
+		expect(log.value).toEqual([0]);
+
+		count = 1;
+		flushSync();
+
+		expect(log.value).toEqual([0, 1]);
+	});
+
+	cleanup();
+});
+```
+
+## E2E tests using Playwright
+
+E2E (short for "end to end") tests allow you to test your full application through the eyes of the user. This section uses [Playwright](https://playwright.dev/) as an example, but you can also use other solutions like [Cypress](https://www.cypress.io/) or [NightwatchJS](https://nightwatchjs.org/).
+
+To get start with Playwright, either let you guide by [their VS Code extension](https://playwright.dev/docs/getting-started-vscode), or install it from the command line using `npm init playwright`. It is also part of the setup CLI when you run `npm create svelte`.
+
+After you've done that, you should have a `tests` folder and a playwright config. You may need to adjust that config to tell Playwright what to do before running the tests - mainly starting your application at a certain port:
+
+```js
+/// file: playwright.config.js
+const config = {
+	webServer: {
+		command: 'npm run build && npm run preview',
+		port: 4173
+	},
+	testDir: 'tests',
+	testMatch: /(.+\.)?(test|spec)\.[jt]s/
+};
+
+export default config;
+```
+
+You can now start writing tests. These are totally unaware of Svelte as a framework, so you mainly interact with the DOM and write assertions.
+
+```js
+/// file: tests/hello-world.spec.js
+import { expect, test } from '@playwright/test';
+
+test('home page has expected h1', async ({ page }) => {
+	await page.goto('/');
+	await expect(page.locator('h1')).toBeVisible();
+});
+```
```

Comment:
I think component testing deserves its own section if it's mentioned at all. I wouldn't recommend installing jsdom above as it's tremendously more complicated than just doing that. You run into all kinds of complications with slots (maybe better with snippets?), events and `use:`, `bind:`, DX of programattically instantiating components, etc. We purposefully avoided setting up component testing in `create-svelte` because the solutions all sucked so badly. I hesitate to recommend anyone component test their Svelte components at the current time. We need to put real effort into making it better and would be doing a disservice to recommend folks use it currently.

```suggestion

## Component testing

Two approaches to component testing are to [use jsdom with vitest](https://github.com/wd-David/svelte-component-test-recipes) or to use [@playwright/experimental-ct-svelte](https://playwright.dev/docs/test-components). `create-svelte` currently does not setup component testing as these approaches are under active development by the community.

```

Reply:
- you can programmatically instantiate slots now using `createRawSnippet`
- what is the problem with events/use/bind?
- "DX of programmatically intantiating components" - what is so bad about it? I mean, yes it's not as nice as declaratively doing that, but doing `mount`/`unmount` isn't so bad? And when it comes to selecting elements in order to interact with them, it's always going to be a bit of legwork (in e2e testing, too)

Reply:
> what is the problem with events/use/bind?

I don't know if it's changed in Svelte 5, but in Svelte 4 there is no programmatic interface to test `bind:`, `use:`, slots, or the context API

> "DX of programmatically intantiating components" - what is so bad about it? I mean, yes it's not as nice as declaratively doing that, but doing mount/unmount isn't so bad? And when it comes to selecting elements in order to interact with them, it's always going to be a bit of legwork (in e2e testing, too)

It's not just instantiating but using features like slots/events/use/bind/context. I can only tell you what the experience was like in Svelte 4 as I haven't seen it in Svelte 5, but I think any docs for component testing would need to show how to do those things.

Reply:
The way to go about all these cases is to create a wrapper component that tests the actual component - I don't see a problem with that. It's how you would have to do this in _any_ framework.
Furthermore, just not documenting how to do component testing because it's cumbersome sounds wrong. Testing components in isolation is something you'd want to do especially when creating a component library, for example. I'm open to having a `> Don't overdo this as component tests are generally harder to maintain` note at the end of the section.
I'll adjust the docs to mention both these things.